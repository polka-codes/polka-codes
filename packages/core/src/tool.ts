export type ToolParameter = {
  name: string
  description: string
  required: boolean
  usageValue?: string
  allowMultiple?: boolean
  children?: ToolParameter[]
}

export type ToolParameterValue = string | { [key: string]: ToolParameterValue } | ToolParameterValue[]

export type ToolExample = {
  description: string
  parameters: { name: string; value: ToolParameterValue }[]
}

export enum PermissionLevel {
  // This tool is completely safe
  None = 0,
  // This tool can read files
  Read = 1,
  // This tool can write files
  Write = 2,
  // This tool can perform arbitrary action
  Arbitrary = 3,
}

export type ToolInfo = {
  name: string
  description: string
  parameters: ToolParameter[]
  examples?: ToolExample[]
  permissionLevel: PermissionLevel
}

export type FullToolInfo = ToolInfo & {
  handler: ToolHandler<ToolInfo, any>
  isAvailable: (provider: any) => boolean
}

export enum ToolResponseType {
  Reply = 'Reply',
  Exit = 'Exit',
  Invalid = 'Invalid',
  Error = 'Error',
  Interrupted = 'Interrupted',
  HandOver = 'HandOver',
  Delegate = 'Delegate',
  Pause = 'Pause',
}

// Reply to the tool use
export type ToolResponseReply = {
  type: ToolResponseType.Reply
  message: string
}

// Should end the message thread
// e.g. task completed successfully
export type ToolResponseExit = {
  type: ToolResponseType.Exit
  message: string
  object?: any
}

// The tool arguments are invalid
export type ToolResponseInvalid = {
  type: ToolResponseType.Invalid
  message: string
}

// Some error occurred when executing the tool
// e.g. network request error, IO error
export type ToolResponseError = {
  type: ToolResponseType.Error
  message: string
  // If true, the tool can be retried
  // e.g. network request error are generally retryable
  // but IO errors are not
  canRetry?: boolean
}

// The tool execution was interrupted
// e.g. user cancelled the tool execution
// or some security policy was violated
export type ToolResponseInterrupted = {
  type: ToolResponseType.Interrupted
  message: string
}

// Hand over the task to another agent
// e.g. hand over a coding task to the coder agent
export type ToolResponseHandOver = {
  type: ToolResponseType.HandOver
  agentName: string
  task: string
  context?: string
  files?: string[]
  originalTask?: string
}

// Delegate the task to another agent
export type ToolResponseDelegate = {
  type: ToolResponseType.Delegate
  agentName: string
  task: string
  context?: string
  files?: string[]
  originalTask?: string
}

export type ToolResponsePause = {
  type: ToolResponseType.Pause
  object: any
}

export type ToolResponse =
  | ToolResponseReply
  | ToolResponseExit
  | ToolResponseInvalid
  | ToolResponseError
  | ToolResponseInterrupted
  | ToolResponseHandOver
  | ToolResponseDelegate
  | ToolResponsePause

export type ToolHandler<T extends ToolInfo, P> = (
  provider: P,
  args: Partial<Record<T['parameters'][number]['name'], ToolParameterValue>>,
) => Promise<ToolResponse>
