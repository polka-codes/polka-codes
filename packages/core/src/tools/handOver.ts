/**
 * Tool for handing over a task to another agent.
 * Generated by polka.codes
 */

import { z } from 'zod'
import { type FullToolInfo, type ToolHandler, type ToolInfo, ToolResponseType } from '../tool'

export const toolInfo = {
  name: 'handOver',
  description: 'Hand over the current task to another agent to complete. This tool MUST NOT to be used with any other tool.',
  parameters: z
    .object({
      agentName: z.string().describe('The name of the agent to hand over the task to').meta({ usageValue: 'Name of the target agent' }),
      task: z.string().describe('The task to be completed by the target agent').meta({ usageValue: 'Task description' }),
      context: z.string().describe('The context information for the task').meta({ usageValue: 'Context information' }),
      files: z
        .preprocess((val) => {
          if (!val) return []
          const values = Array.isArray(val) ? val : [val]
          return values.flatMap((i) => (typeof i === 'string' ? i.split(',') : [])).filter((s) => s.length > 0)
        }, z.array(z.string()))
        .optional()
        .describe('The files relevant to the task. Comma separated paths')
        .meta({ usageValue: 'Relevant files' }),
    })
    .meta({
      examples: [
        {
          description: 'Hand over a coding task to the coder agent',
          input: {
            agentName: 'coder',
            task: 'Implement the login feature',
            context: 'We need a secure login system with email and password',
            files: 'src/auth/login.ts,src/auth/types.ts',
          },
        },
      ],
    }),
} as const satisfies ToolInfo

export const handler: ToolHandler<typeof toolInfo, any> = async (_provider, args) => {
  const parsed = toolInfo.parameters.safeParse(args)
  if (!parsed.success) {
    return {
      type: ToolResponseType.Invalid,
      message: {
        type: 'error-text',
        value: `Invalid arguments for handOver: ${parsed.error.message}`,
      },
    }
  }
  const { agentName, task, context, files } = parsed.data

  return {
    type: ToolResponseType.HandOver,
    agentName,
    task,
    context,
    files: files ?? [],
  }
}

export const isAvailable = (_provider: any): boolean => {
  return true
}

export default {
  ...toolInfo,
  handler,
  isAvailable,
} satisfies FullToolInfo
