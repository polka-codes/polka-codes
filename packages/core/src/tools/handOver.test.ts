/**
 * Tests for the handOver tool
 * Generated by polka.codes
 */

import { describe, expect, test } from 'bun:test'
import { ToolResponseType } from '../tool'
import { handler, toolInfo } from './handOver'

describe('handOver tool', () => {
  test('tool info should be properly defined', () => {
    expect(toolInfo.name).toBe('hand_over')
    expect(toolInfo.parameters).toHaveLength(4)
    expect(toolInfo.examples).toHaveLength(1)
  })

  test('handler should return proper handover response', async () => {
    const result = await handler(
      {},
      {
        agent_name: 'coder',
        task: 'implement feature',
        context: 'test context',
        files: 'file1.ts,file2.ts',
      },
    )

    expect(result).toEqual({
      type: ToolResponseType.HandOver,
      agentName: 'coder',
      task: 'implement feature',
      context: 'test context',
      files: ['file1.ts', 'file2.ts'],
    })
  })

  test('handler should throw error if required parameters are missing', async () => {
    await expect(handler({}, {})).rejects.toThrow()
  })
})
