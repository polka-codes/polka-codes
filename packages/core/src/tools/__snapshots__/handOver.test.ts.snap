// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`handOver tool handler should return proper handover response 1`] = `
{
  "agentName": "coder",
  "context": "test context",
  "files": [
    "file1.ts",
    "file2.ts",
  ],
  "task": "implement feature",
  "type": "HandOver",
}
`;
